programs/address-map/src/instruction.rs

```diff
@@ -0,0 +1,177 @@
+use {
+    crate::id,
+    serde::{Deserialize, Serialize},
+    solana_sdk::{
+        clock::Epoch,
+        instruction::{AccountMeta, Instruction},
+        pubkey::Pubkey,
+        system_program,
+    },
+};
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq, Clone)]
+pub enum AddressMapInstruction {
+    /// Initialize an address map account
+    ///
+    /// # Account references
+    ///   0. `[WRITE]` Uninitialized address map account
+    ///   1. `[SIGNER, WRITE]` Account that will fund the new address map.
+    ///   2. `[]` System program for CPI.
+    InitializeAccount {
+        bump_seed: u8,
+        num_entries: u8,
+        authority: Pubkey,
+    },
+
+    /// Set a new authority for an address map account
+    ///
+    /// # Account references
+    ///   0. `[WRITE]` Address map account to set authority for
+    ///   1. `[SIGNER]` Current authority
+    SetAuthority { new_authority: Option<Pubkey> },
+
+    /// Insert entries into an address map account
+    ///
+    /// # Account references
+    ///   0. `[WRITE]` Address map account to insert entries into
+    ///   1. `[SIGNER]` Current authority
+    InsertEntries { offset: u8, entries: Vec<Pubkey> },
+
+    /// Activate an address map account
+    ///
+    /// # Account references
+    ///   0. `[WRITE]` Address map account to activate
+    ///   1. `[SIGNER]` Current authority
+    Activate,
+
+    /// Deactivate an address map account
+    ///
+    /// # Account references
+    ///   0. `[WRITE]` Address map account to deactivate
+    ///   1. `[SIGNER]` Current authority
+    Deactivate,
+
+    /// Close an address map account
+    ///
+    /// # Account references
+    ///   0. `[WRITE]` Address map account to close
+    ///   1. `[WRITE]` Recipient of closed account lamports
+    ///   2. `[SIGNER]` Current authority, if required.
+    CloseAccount,
+}
+
+/// Derive an address map address from a wallet address and the current epoch.
+pub fn derive_address_map_address(payer_address: &Pubkey, current_epoch: Epoch) -> (Pubkey, u8) {
```

Comment:
I'm worried that this will make discoverability of address maps difficult.  One of the great features of the associated token account is that you can immediately find someone's token account with just their wallet and the mint.  Here, you may have to iterate through a bunch of epochs in order to find their address map, or you'll just use `find_program_address` and do it the "easy" way.

What if, instead of an epoch, this is another `Pubkey`? That way, you can setup a different address map for each program that you interact with.  Since you've cleverly put this on a PDA dependent on the signer, there's no way for one of these to get hijacked after `CloseAccount`.

Reply:
I agree that it's not ideal and I did consider this point when reimplementing the program here: https://github.com/solana-labs/solana/pull/21616. However, I think indexing or creating a known registry of on-chain lookup tables is going to have to be the way forward here. I think it's more likely that bots and protocols will manage their own lookup table accounts and users will never have to manage them.